### ISSUE TITLE
Remove subcats of Category:Draft articles when accepting

### ISSUE THREAD

**Issue by enterprisey** [WPAFC/afch-rewrite#131]
https://en.wikipedia.org/w/index.php?diff=1001040898
davidwr at 23:59, 17 January 2021 (UTC)

**Comment by satcasm** [comment]
Hi @enterprisey , would like to take this as my next issue . Hope you could brief me about it, I went through the links you provided couldn't make much out of it.
Much appreciated!

**Comment by enterprisey** [comment]
@satcasm So there are two changes. Let's call all subcategories of [Category:Draft articles](https://en.wikipedia.org/wiki/Category:Draft_articles) "draft categories". Then these "draft categories" should:
1. not be "disabled" during the cleaning process (in the "else" block of the isAccept if statement in AFCH.Text.prototype.cleanUp)
2. be removed during accepting (I would recommend skipping them in the $.each block in AFCH.Text.prototype.updateCategories)

I can think of many ways to check if a category is a draft category. The "cheating" way is to just check if the name of the draft starts with "Drafts about" (or is equal to "Promising drafts"), because that catches most of them, but that wouldn't handle most of the subcategories. The better way, which we should use in the final product (maybe just use the "cheating" way to check that your category handling code above is correct, temporarily) is to use localStorage, like the loadWikiProjectList function. We could make a getDraftArticlesSubcats function that does the same thing (if lsKey is in localStorage, parse and return that array of categories; otherwise, get it from the API and then return it). You'll probably need https://www.mediawiki.org/wiki/API:Categorymembers, with cmtype=subcat. Unfortunately, it doesn't return subcategories of subcategories, so you'll need recursion... perhaps write a recursive getSubcats function, and then call it from getDraftArticlesSubcats with the parameter "Draft articles". Recursion and callbacks do not mix well at all; feel free to let me know when you get to that part.

So, once you have getDraftArticlesSubcats, you'll call it from the two places I checked above and then check if every category is in the return value, doing the appropriate thing in each place. Let me know what I should clarify or if you have any questions.

**Comment by satcasm** [comment]
@enterprisey  in the else part of AFCH.Text.prototype.cleanUp nothing is mentioned about subcategories. So what exactly did you mean by ```not be "disabled" ```?


![image](https://user-images.githubusercontent.com/56315878/108635118-81bbec80-74a3-11eb-8ca3-1c5320a21a88.png)

**Comment by enterprisey** [comment]
The comment is inaccurate; changing `[[Category:Foo]]` to `[[:Category:Foo]]` is "disabling" it (or to be completely accurate, "changing it to a link"). So I mean that this line of code should change `[[Category:Rocks]]` to `[[:Category:Rocks]]`, but it should not change `[[Category:Drafts about people]]` to `[[:Category:Drafts about people]]` because Category:Drafts about people is a subcategory of Category:Draft articles.

**Comment by satcasm** [comment]
Hey @enterprisey , so I started small with the "cheating method".
This is my code of updateCategories:
![image](https://user-images.githubusercontent.com/56315878/108666160-82cd3800-74fc-11eb-87cb-4472933c671b.png)
This is of else part:
![image](https://user-images.githubusercontent.com/56315878/108666200-98426200-74fc-11eb-9244-fea3ea2c39fb.png)
Can you tell what next step would be? I mean how to check it?

**Comment by enterprisey** [comment]
In both of those functions, `text` is the whole text of the page; you may want to pass a function as the second parameter to `replace` in cleanUp.

**Comment by satcasm** [comment]
> In both of those functions, `text` is the whole text of the page; you may want to pass a function as the second parameter to `replace` in cleanUp.

But for the time being is there something wrong with this code ? I mean I am using ```.test``` which should work just fine. Also how to get into the else part? If I accept an article it's just going into ```if(isAccept)``` loop.

**Comment by enterprisey** [comment]
Yeah, if you use `text` then if you have two categories (and only one is a subcategory of Category:Draft articles), the code will act as if both are subcategories, which would be buggy. To use the else part, try using "clean", which you can get to by opening the "<<" menu on the far right side of the main menu.

**Comment by satcasm** [comment]
Okay so what should I pass along with the ```text``` as the syntax of categories and subcategories is similiar?

**Comment by enterprisey** [comment]
I would recommend creating a helper function, getDraftArticlesSubcats, that returns a list of categories; for now, you could just have it return a hard-coded array with a single item for testing, like Category:Drafts about people. Then, in updateCategories, check if `category` is in that returned array, and return from the callback if true. And in cleanUp, make the second parameter of .replace a callback, and check if the matched text is in that returned array.
